packages/rest-crud/README.md

```diff
@@ -19,25 +19,20 @@ npm install --save @loopback/rest-crud
 
 2. Create a Repository class, e.g. using `lb4 repository` tool.
 
-3. Create a base REST CRUD controller class for your model.
+3. Create a REST CRUD controller class for your model.
 
    ```ts
-   const CrudRestController = defineCrudRestController<
+   const ProductController = defineCrudRestController<
      Product,
      typeof Product.prototype.id,
      'id'
    >(Product, {basePath: '/products'});
    ```
 
-4. Create a new subclass of the base controller class to configure repository
-   injection.
+4. Inject the model's repository via Dependency Injection.
 
    ```ts
-   class ProductController extends CrudRestController {
-     constructor(@repository(ProductRepository) repo: ProductRepository) {
-       super(repo);
-     }
-   }
+   inject('repositories.ProductRepository')(ProductController, undefined, 0);
```

Comment:
Why not using the subclassing for dependency injection? It seems to be easier to understand.

Reply:
> It seems to be easier to understand.

Agreed. However, there some redundancy going on now.

Earlier `defineCrudRestController` used to return a `CrudRestControllerImpl` class which was assigned to `CrudRestController`, so it made sense to do `class ProductController extends CrudRestController`.

Now `defineCrudRestController` returns a named controller class like `ProductController`. Doing `class MyProductController extends ProductController`, looks weird. "Why can't I just use `ProductController`?"

Alternative would something like:
```
const ProductControllerBase = defineCrudRestController<
     Product,
     typeof Product.prototype.id,
     'id'
>(Product, {basePath: '/products'});

class ProductController extends ProductControllerBase {}
```

@strongloop/loopback-next thoughts?

Reply:
The problem with subclassing is naming - the class created by `defineCrudRestController` already has the right name (`ProductController`). How should the new class be called? Should we use an anonymous class?

```ts
const controller = const ProductController = defineCrudRestController<
  Product,
  typeof Product.prototype.id,
  'id'
>(Product, {basePath: '/products'});

const ProductController = class extends controller {
  constructor(
    // args decorated with @inject
  ) {
    super(/*forward args*/);
  }
}
```

I am also concerned about micro-performance - by creating a new class, we are adding more code to be executed for every incoming request.

Anyhow, these instructions will soon be moved to "Advanced use", most LB users are going to use ModelApiBooter that will be introduced by one of the next tasks in this Epic.

Reply:
If we always expect to inject a repository, we can either make the repository class an argument of `defineCrudRestController()` or create a helper function such as:

```ts
setupRepositoryInjection(ProductController, 'repositories.ProductRepository');
```

Reply:
I am in favor of making the repository class an argument of `defineCrudRestController` makes sense to me. @bajtos thoughts?

Reply:
@raymondfeng I experimented the "repository class argument" approach:

1. Passing a binding key would have been best, but since we have no access to the context, it won't work.
2. Passing a repository instance requires the repository to be setup with datasource. It brings in the datasource factor, meaning more work and complexity.

We don't always expect to inject a repository, though. Eg: `class ProductController extends ProductControllerBase {}`.

Reply:
I am neutral about `setupRepositoryInjection`, will go with @bajtos's opinion.

Reply:
@hacksparrow Two points:

1. No context is needed. We only require the binding key or class for the repository so that we can have:

```ts
export function defineCrudRestController<
  T extends Entity,
  IdType,
  IdName extends keyof T,
  R extends EntityCrudRepository<T, IdType, Relations>,
  Relations extends object = {}
>(
  modelCtor: typeof Entity & {prototype: T & {[key in IdName]: IdType}},
  repoCtor: R,
  options: CrudRestControllerOptions,
): CrudRestControllerCtor<T, IdType, IdName, Relations> {
  // ...
  @api({basePath: options.basePath, paths: {}})
  class CrudRestControllerImpl
    implements CrudRestController<T, IdType, IdName> {
    constructor(
      public readonly @repository(ProductRepository): R,
    ) {}
```

2. Without the repository injection setup, the controller class is not functional.

Reply:
I am strongly against setting up DI in `defineCrudRestController`. It's a code smell when DI-related constructs are littered across the code base. The class created by `defineCrudRestController` should be usable without DI too, it is the responsibility of the caller to decide how they want to supply the needed repository instance.

In some cases, the user may not want to inject the repository via DI, they may want to receive the datasource instance via DI instead and create a Repository instance on the fly, for example by calling `new DefaultCrudRepository(Product, ds)`.

IMO, this discussion has uncovered a major design limitation of our IoC container, where it's not possible to wire up dependencies from outside. In a well-designed app, it's the app that defines how are dependencies resolved, not the individual classes!

Ideally, I would like to wire my IoC setup along the following lines (a low-level example, we would want a nicer API in a real implementation):

```ts
const ProductController = defineCrudRestController(...);
const binding = app.controller(ProductController);
binding.injectConstructorArg(0, 'repositories.ProductRepository');
```

Reply:
I am proposing to land the current version and open a new issue or a new pull request to discuss how to improve the example code snippet.

As I mentioned above, this code snippet will be used by a very small fraction of our users, let's invest our time into improvements with bigger impact.